File: Leetcode/904 - Fruit Into Baskets/solution.cpp
int totalFruit(vector<int>& fruits) {
    int low = 0, high = 0;
    int maxFruits = 0;
    unordered_map<int, int> fruitCount;
    while (high < fruits.size()) {
        fruitCount[fruits[high]]++;
        while (fruitCount.size() > 2) {
            fruitCount[fruits[low]]--;
            if (fruitCount[fruits[low]] == 0) fruitCount.erase(fruits[low]);
            low++;
        }
        maxFruits = max(maxFruits, high - low + 1);
        high++;
    }
    return maxFruits;
}

File: Leetcode/904 - Fruit Into Baskets/solution.js
var totalFruit = function(fruits) {
    let low = 0;
    let maxFruits = 0;
    const fruitCount = new Map();

    for (let high = 0; high < fruits.length; high++) {
        const fruit = fruits[high];
        fruitCount.set(fruit, (fruitCount.get(fruit) || 0) + 1);

        while (fruitCount.size > 2) {
            const leftFruit = fruits[low];
            const count = fruitCount.get(leftFruit) - 1;
            if (count === 0) {
                fruitCount.delete(leftFruit);
            } else {
                fruitCount.set(leftFruit, count);
            }
            low++;
        }
        maxFruits = Math.max(maxFruits, high - low + 1);
    }

    return maxFruits;
};